feat: EmailVerificationQueries migration to Hibernate

anugrahsinghal opened the pull request:
- [x] create domain models
- [x] modify queries
- [x] additional error handling in Start.java
- [x] modify changelog
- [x] all tests should pass
- [x] Entity Null Checks when doing `session.get`
- [x] verify data flow to ensure `session.get` instead of HQL

anugrahsinghal committed:
feat: domain model for EmailVerificationQueries

anugrahsinghal committed:
feat: migrate delete queries

anugrahsinghal committed:
feat: add entity mapping for email-verification

anugrahsinghal committed:
refactor: transactional query gets session directly

anugrahsinghal committed:
feat: add allArgsConstructor

anugrahsinghal committed:
feat: migrate insert queries to use hibernate

anugrahsinghal committed:
feat: migrate select queries to use hibernate

anugrahsinghal committed:
feat: handling for PersistenceException

anugrahsinghal commented:
**Note**: draft due to failing tests in Start.java

anugrahsinghal committed:
docs: add changelog

anugrahsinghal committed:
refactor: change variable name

anugrahsinghal committed:
feat: add handling for PersistenceException

anugrahsinghal commented:
hey @rishabhpoddar 

for test `io.supertokens.test.emailverification.EmailVerificationTest:testVerifyingSameEmailTwice`

the test fails at the following statement

```java

StorageLayer.getEmailVerificationStorage(process.getProcess()).startTransaction(con -> {
// this throws an exception which is to be expected
// as email is already verified
// earlier it would be swallowed by intention
// see io.supertokens.storage.sql.Start:L701 --> we do not throw an error since the email is already verified

StorageLayer.getEmailVerificationStorage(process.getProcess())
.updateIsEmailVerified_Transaction(con, user.id, user.email, true);

            return null;
        });
// but now the exception is thrown when doing a transaction.commit()
// hence error is at StorageLayer.getEmailVerificationStorage(process.getProcess()).startTransaction(...)
// how do you propose we manage this ?

// managing EmailVerification Specific scenario at startTransaction(...)
// does not seem like a good idea, as the function serves a generic purpose
```

have added my comments, let me know your thoughts on this

anugrahsinghal committed:
chore: remove unnecessary comment

anugrahsinghal committed:
docs: add raw sql query

rishabhpoddar commented:
## Options:
- Use save() -> flush() instead of just save().
   - We should follow the same pattern for all transaction based updates.
   - We will have to change [this](https://github.com/anugrahsinghal/supertokens-sql-plugin/blob/feature%2Femail-verification/src/main/java/io/supertokens/storage/sql/hibernate/CustomSessionWrapper.java#L287) to not throw (or make our own saveAndFlush())
   - Is there any disadvantage to calling flush ourselves? For example, does it clear the level 1 cache of the session?

- Other option is to extend TransactionConnection to allow callbacks to handle error post commit:
```
con

st => {
    q1

    q2 -> con.addPersistentErrorHandler((err) => {
        if (err.message.contains("m1")) {
            throw CustomError();
        }
        // else ignore
    }) (e1)

    q3 -> con.addPersistentErrorHandler((err) => {
        if (err.message.contains("m2")) {
            throw CustomError1();
        }
        // else ignore
    }) (e2)

}

commit() => {}
```

anugrahsinghal committed:
feat: use session.flush with update query

when doing double calls to the same method
exceptions are not raised until the transactions are committed
to avoid this issue, we use session.flush to ensure consistency
b/w DB and Hibernate model

anugrahsinghal committed:
feat: add error handling for hibernate

anugrahsinghal committed:
fixL: change args according param type changes

anugrahsinghal committed:
test: change test to use different transactions

anugrahsinghal committed:
chore: remove @Deprecated

anugrahsinghal commented on src/test/java/io/supertokens/storage/sql/test/ExceptionParsingTest.java:
```diff
@@ -213,12 +213,18 @@ public void updateIsEmailVerified_TransactionExceptions()
-                storage.updateIsEmailVerified_Transaction(conn, userId, userEmail, true);
-                return true;
-            });
+            {
+                storage.startTransaction(conn -> {
+                    storage.updateIsEmailVerified_Transaction(conn, userId, userEmail, true);
+                    return true;
+                });
+
+                storage.startTransaction(conn -> {
+                    storage.updateIsEmailVerified_Transaction(conn, userId, userEmail, true);
+                    // The insert in this call throws, but it's swallowed in the method
+                    return true;
+                });
+            }
```
@rishabhpoddar 

changed this test according to our discussion.

the same function called in two different transaction

anugrahsinghal commented on src/main/java/io/supertokens/storage/sql/Start.java:
```diff
@@ -209,19 +209,19 @@ public <T> T startTransaction(TransactionLogic<T> logic, TransactionIsolationLev
+                if (e instanceof PersistenceException) {
+                    actualException = e.getCause();
+                } else if (e instanceof StorageQueryException) {
                     actualException = e.getCause();
                 } else if (e instanceof StorageTransactionLogicException) {
                     actualException = ((StorageTransactionLogicException) e).actualException;
-                } else if (e instanceof LockAcquisitionException) {
+                }
+
+                if (actualException instanceof LockAcquisitionException) {
                     // LockAcquisitionException -> PSQLException
-                    actualException = e.getCause();
-                } else if (e instanceof OptimisticLockException) {
-                    // OptimisticLockException -> LockAcquisitionException -> PSQLException
-                    actualException = e.getCause().getCause();
+                    actualException = actualException.getCause();
```
I think it would be better to revert/comment out this logic for now.

We can handle this in follow-up commits.

WDYT ?

rishabhpoddar replied:
Yea sure. We can comment it out

anugrahsinghal replied:
done

anugrahsinghal committed:
revert: exception handling with Server message

anugrahsinghal committed:
fix: reapply deprecated annotation to unused func

anugrahsinghal committed:
style: move method for consistent diff

anugrahsinghal committed:
style: minor string concat changes

anugrahsinghal committed:
revert: Exception Handling logic

anugrahsinghal committed:
style: apply spotless plugin

anugrahsinghal committed:
fix: correct method chaining

anugrahsinghal committed:
refactor: use session.get for entity
